File: python/agent/agent_basic.py
import os
from llm import LLMInterface

# Configuration
LLM_PROVIDER = "openai"  # Can be "openai" or "anthropic"

SYS_PROMPT = """
You are a helpful agent that can read files and list directory contents. 
You have access to two tools:
1. list_files - to list files in a directory
2. read_file - to read the contents of a file

Use these tools when the user asks questions about files or directories.
"""


def list_files(path: str = ".") -> str:
    """List files and directories in the given path"""
    try:
        if not os.path.exists(path):
            return f"Error: Path '{path}' does not exist"

        items = os.listdir(path)
        if not items:
            return f"Directory '{path}' is empty"

        files = []
        dirs = []

        for item in items:
            item_path = os.path.join(path, item)
            if os.path.isdir(item_path):
                dirs.append(f"📁 {item}/")
            else:
                files.append(f"📄 {item}")

        result = f"Contents of '{path}':\n"
        if dirs:
            result += "\nDirectories:\n" + "\n".join(dirs)
        if files:
            result += "\nFiles:\n" + "\n".join(files)

        return result
    except Exception as e:
        return f"Error listing files: {str(e)}"


def read_file_content(filepath: str) -> str:
    """Read the contents of a file"""
    try:
        if not os.path.exists(filepath):
            return f"Error: File '{filepath}' does not exist"

        if os.path.isdir(filepath):
            return f"Error: '{filepath}' is a directory, not a file"

        with open(filepath, "r", encoding="utf-8") as f:
            content = f.read()

        return f"Contents of '{filepath}':\n```\n{content}\n```"
    except Exception as e:
        return f"Error reading file: {str(e)}"


def execute_tool(tool_call: dict, llm: LLMInterface) -> str:
    """Execute a tool call and return the result"""
    tool_name = tool_call["name"]
    tool_args = tool_call["args"]

    print(f"🔧 Calling tool: {tool_name} with args: {tool_args}")

    if tool_name == "list_files":
        path = tool_args.get("path", ".")
        return llm.list_files_filtered(path)
    elif tool_name == "read_file":
        filepath = tool_args.get("filepath", "")
        return llm.read_file_filtered(filepath)
    else:
        return f"Error: Unknown tool '{tool_name}'"


def run_agent():
    """Main agent loop"""
    print(f"🤖 File Agent (using {LLM_PROVIDER.upper()}) - Ready to help!")
    print("Type 'quit' to exit")
    print("-" * 50)

    # Initialize LLM interface
    llm = LLMInterface(LLM_PROVIDER)

    # Initialize conversation
    messages = []
    waiting_for_user_input = True

    while True:
        if waiting_for_user_input:
            user_input = input("\n💬 You: ").strip()

            if user_input.lower() in ["quit", "exit", "q"]:
                print("👋 Goodbye!")
                break

            if not user_input:
                continue

            messages.append({"role": "user", "content": user_input})
            waiting_for_user_input = False

        print("\n🤖 Agent: ", end="", flush=True)

        # Get completion from LLM
        response_text, tool_calls = llm.create_completion(messages, SYS_PROMPT)

        if tool_calls:
            # Add assistant message with tool calls for OpenAI compatibility
            llm.add_assistant_message_with_tools(messages, tool_calls)

            # Execute each tool call
            for tool_call in tool_calls:
                result = execute_tool(tool_call, llm)
                llm.add_tool_response(messages, tool_call, result)

            # Continue the loop to get LLM response after tool calls
            continue
        else:
            # No tool calls, print the final response
            print(response_text)
            messages.append({"role": "assistant", "content": response_text})
            waiting_for_user_input = True


if __name__ == "__main__":
    run_agent()

File: node/src/agent/agent_basic.ts
import * as readline from 'readline';
import { LLMInterface } from './llm';

// Configuration
const LLM_PROVIDER = 'anthropic'; // Can be "openai" or "anthropic"

const SYS_PROMPT = `
You are a helpful agent that can read files and list directory contents. 
You have access to two tools:
1. list_files - to list files in a directory
2. read_file - to read the contents of a file

Use these tools when the user asks questions about files or directories.
`;

interface Message {
  role: 'user' | 'assistant' | 'system' | 'tool';
  content: string | any[];
  tool_calls?: any[];
  tool_call_id?: string;
  name?: string;
}

interface ToolCall {
  id: string;
  name: string;
  args: Record<string, any>;
}

function executeTool(toolCall: ToolCall, llm: LLMInterface): string {
  const toolName = toolCall.name;
  const toolArgs = toolCall.args;

  console.log(`🔧 Calling tool: ${toolName} with args:`, toolArgs);

  if (toolName === 'list_files') {
    const path = toolArgs.path || '.';
    return llm.listFilesFiltered(path);
  } else if (toolName === 'read_file') {
    const filepath = toolArgs.filepath || '';
    return llm.readFileFiltered(filepath);
  } else {
    return `Error: Unknown tool '${toolName}'`;
  }
}

async function runAgent(): Promise<void> {
  console.log(`🤖 File Agent (using ${LLM_PROVIDER.toUpperCase()}) - Ready to help!`);
  console.log("Type 'quit' to exit");
  console.log('-'.repeat(50));

  // Initialize LLM interface
  const llm = new LLMInterface(LLM_PROVIDER);

  // Initialize conversation
  const messages: Message[] = [];
  let waitingForUserInput = true;

  // Setup readline interface
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const askQuestion = (query: string): Promise<string> => {
    return new Promise((resolve) => {
      rl.question(query, (answer) => {
        resolve(answer);
      });
    });
  };

  try {
    while (true) {
      if (waitingForUserInput) {
        const userInput = await askQuestion('\n💬 You: ');

        if (['quit', 'exit', 'q'].includes(userInput.toLowerCase().trim())) {
          console.log('👋 Goodbye!');
          break;
        }

        if (!userInput.trim()) {
          continue;
        }

        messages.push({ role: 'user', content: userInput.trim() });
        waitingForUserInput = false;
      }

      process.stdout.write('\n🤖 Agent: ');

      // Get completion from LLM
      const [responseText, toolCalls] = await llm.createCompletion(messages, SYS_PROMPT);

      if (toolCalls) {
        // Add assistant message with tool calls for OpenAI compatibility
        llm.addAssistantMessageWithTools(messages, toolCalls);

        // Execute each tool call
        for (const toolCall of toolCalls) {
          const result = executeTool(toolCall, llm);
          llm.addToolResponse(messages, toolCall, result);
        }

        // Continue the loop to get LLM response after tool calls
        continue;
      } else {
        // No tool calls, print the final response
        console.log(responseText);
        messages.push({ role: 'assistant', content: responseText || '' });
        waitingForUserInput = true;
      }
    }
  } catch (error) {
    console.error('Error:', error);
  } finally {
    rl.close();
  }
}

// Only run if this file is executed directly
if (require.main === module) {
  runAgent().catch(console.error);
}

export { runAgent }; 